File: min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.ls = []
        self.minls = []

    def push(self, val: int) -> None:
        self.ls.append(val)
        if not self.minls or val <= self.minls[-1]:
            self.minls.append(val)

    def pop(self) -> None:
        val = self.ls.pop()
        if val == self.minls[-1]:
            self.minls.pop()
            
    def top(self) -> int:
        return self.ls[-1] if self.ls else None

    def getMin(self) -> int:
        return self.minls[-1] if self.minls else None

File: min-stack/min-stack.cpp
class MinStack {
public:
    stack<int> normal;
    stack<int> minStack;
    MinStack() {
        
    }
    
    void push(int val) {
        if (minStack.empty() || val <= minStack.top()) {
            minStack.push(val);
            normal.push(val);
        } else {
            normal.push(val);
        }
    }
    
    void pop() {
        if (normal.top() == minStack.top()) {
            minStack.pop();
            normal.pop();
        } else {
            normal.pop();
        }
    }
    
    int top() {
        return normal.top(); 
    }
    
    int getMin() {
        return minStack.top(); 
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
